File: 3819-count-covered-buildings/3819-count-covered-buildings.java
class Solution {
    public int countCoveredBuildings(int n, int[][] nums) {
        int count = 0;
        Map<Integer, TreeSet<Integer>> mapx = new HashMap<>();
        Map<Integer, TreeSet<Integer>> mapy = new HashMap<>();
        
        for(int i = 0; i < nums.length; i++){
            int x = nums[i][0];
            int y = nums[i][1];
            mapx.putIfAbsent(x, new TreeSet<>());
            mapy.putIfAbsent(y, new TreeSet<>());
            mapx.get(x).add(y);
            mapy.get(y).add(x);
        }
        
        for(int i = 0; i < nums.length; i++){
            int x = nums[i][0];
            int y = nums[i][1];
            
            //left right
            if(mapy.get(y).first() >= x || mapy.get(y).last() <= x) continue;
            
            //top bottom
            if(mapx.get(x).first() >= y || mapx.get(x).last() <= y) continue;
            
            count++;
        }
        return count;
    }
}

File: 3819-count-covered-buildings/3819-count-covered-buildings.cpp
class Solution {
public:
    int countCoveredBuildings(int n, vector<vector<int>>& nums) {
        int count = 0;
        unordered_map<int,set<int>> mapx;
        unordered_map<int,set<int>> mapy;
        for(int i = 0 ; i < nums.size() ; i++){
            int x = nums[i][0];
            int y= nums[i][1];
            mapx[x].insert(y);
            mapy[y].insert(x);
        }
        for(int i = 0 ; i < nums.size() ; i++){
           int x = nums[i][0];
           int y = nums[i][1];
           //left right
           if(*mapy[y].begin()>=x || *mapy[y].rbegin()<=x) continue;

           //top bottom
           if(*mapx[x].begin()>=y || *mapx[x].rbegin()<=y) continue;

           count++;
        }
        return count;
    }
};